miden/src/tools/mod.rs

```diff
@@ -0,0 +1,61 @@
+use assembly::Assembler;
+use vm_core::{Operation, ProgramInputs};
+
+#[derive(Debug, PartialEq)]
+pub struct ProgramInfo {
+    total_vm_cycles: usize,
+    total_noops: usize,
+}
+
+impl ProgramInfo {
+    /// Creates a new ProgramInfo object
+    pub fn new(total_vm_cycles: usize, total_noops: usize) -> ProgramInfo {
+        Self {
+            total_vm_cycles,
+            total_noops,
+        }
+    }
+
+    /// Returns total vm cycles to execute a program
+    pub fn total_vm_cycles(&self) -> usize {
+        self.total_vm_cycles
+    }
+
+    /// Returns total noops executed as part of a program
+    pub fn total_noops(&self) -> usize {
+        self.total_noops
+    }
+}
+
+/// Returns program analysis for a given script string. Returns ProgramInfo with following fields:
+/// - total_vm_cycles (vm cycles it takes to execute the entire script)
+/// - total_noops (total noops executed as part of a program)
+pub fn analyze(program: &str, inputs: ProgramInputs) -> ProgramInfo {
+    let assembler = Assembler::new();
+    let script = assembler.compile_script(program).unwrap();
+    let vm_state_iterator = processor::execute_iter(&script, &inputs);
+
+    let mut total_vm_cycles = 0;
+    let mut noop_count = 0;
+
+    for vm_state in vm_state_iterator {
+        if matches!(vm_state.as_ref().unwrap().op, Some(Operation::Noop)) {
```

Comment:
Do you need the `unwrap()` here? Would something like this work: `matches!(vm_state.op, Some(Operation::Noop))`?

Reply:
Oh - I think it won't. I'm trying to think of a way to get rid of the `unwrap`'s here - ideally, we shouldn't have them in production code.

Reply:
I guess one thing that this doesn't account for is when a program terminates with an error. We should probably change the return type of `analyze()` to return a `Result<ProgramInfo, ExecutionError>` and then if we encounter an error it would be just returned instead of `ProgramInfo`. What do you think?

Reply:
> I'm trying to think of a way to get rid of the unwrap's here - ideally, we shouldn't have them in production code.
Just to clarify, the reason is proper error propagation right?
Can we use something like:
```
for state in vm_state_iterator {
        let vm_state = match state {
            Ok(vm_state) => vm_state,
            Err(err) => return Err(err)
        };
        if matches!(vm_state.op, Some(Operation::Noop)) {
            noop_count += 1;
        }
        total_vm_cycles = vm_state.clk;
    }
```

Reply:
> I guess one thing that this doesn't account for is when a program terminates with an error. We should probably change the return type of analyze() to return a Result<ProgramInfo, ExecutionError> and then if we encounter an error it would be just returned instead of ProgramInfo. What do you think?

Makes sense to propagate error here. Since there are two possible errors here, `AssemblyError` while compiling the script and `ExecutionError` while executing the script, should we add a new enum called `AnalysisError` and return that instead? That doesn't look very clean though. Is there a better (more idiomatic) way to handle that? Thanks.

Reply:
Yes - the reason is proper error propagation, but also, as a general rule, I try not to have `unwrap`'s in non-test code.

Yes, let's add a new enum. We did something like this [here](https://github.com/maticnetwork/miden/blob/next/miden/tests/integration/helpers/mod.rs#L19). Though I don't love the name `AnalysisError`. One other option that came to mind is `MidenError` - but I don't love this one too much either.

The code snippet above looks right - though, I'd prefer to use [map_err()](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err) instead of `match`. For example, something like:
```Rust
for state in vm_state_iterator {
    let vm_state = state.map_err(...)?;
    if matches!(vm_state.op, Some(Operation::Noop)) {
        noop_count += 1;
    }
    total_vm_cycles = vm_state.clk;
}
```

Reply:
Maybe `ProgramError` is better than the other options?

Reply:
@bobbinth - `ProgramError` sounds good to me. In general, do you prefer more verbose names like if this enum is only going to be used for program analysis, then `ProgramAnalysisError` would be more verbose but not very clean or should we lean more towards succinctness.